bindings/java/src/main/native/jucx_common_def.cc

```diff
@@ -189,12 +191,26 @@ UCS_PROFILE_FUNC_VOID(jucx_request_callback, (request, status), void *request, u
     env->DeleteGlobalRef(ctx->jucx_request);
     ctx->callback = NULL;
     ctx->jucx_request = NULL;
+    ctx->is_completed = false;
     ucp_request_free(request);
 }
 
 void recv_callback(void *request, ucs_status_t status, ucp_tag_recv_info_t *info)
 {
-    jucx_request_callback(request, status);
+    struct jucx_context *ctx = (struct jucx_context *)request;
+    if (ctx->jucx_request != NULL) {
```

Comment:
the lock scope does not contain the check "(ctx->jucx_request != NULL)" so it means it probably not really thread safe
also, we should be able to remove the sched_yield() loop and memory barriers now that we use a lock. it will cleanup the code and make thread safety easier to prove/disprove